### PR TITLE
lambda.part2.exercise.Exercise2(Buchok)

### PR DESCRIPTION
**Pull request by Boochok**

### CODE REVIEW

File: src/test/java/lambda/part2/exercise/Exercise2.java [hunk 1]
```diff
@@ -25,75 +25,91 @@ public void personHasNotEmptyLastNameAndFirstName() {
...
+        assertTrue(personHasEmptyFirstName.test(new Person("", "Мельников", 20)));
+        assertTrue(personHasEmptyLastName.test(new Person("", "", 20)));
     }
 
     // TODO метод (T -> boolean) -> (T -> boolean)
     // TODO - возвращает новый предикат, являющийся отрицанием исходного
     // TODO - при реализации использовать логический оператор !
     private <T> Predicate<T> negate(Predicate<T> predicate) {
-        throw new UnsupportedOperationException();
+        return predicate.negate();
```
Comment (jemika) [inline src/test/java/lambda/part2/exercise/Exercise2.java]: read TODO description ( при реализации использовать логический оператор !)

File: src/test/java/lambda/part2/exercise/Exercise2.java [hunk 2]
```diff
@@ -25,75 +25,91 @@ public void personHasNotEmptyLastNameAndFirstName() {
...
     }
 
     // TODO метод (T -> boolean, T -> boolean) -> (T -> boolean)
     // TODO - возвращает новый предикат, объединяющий исходные с помощью операции "AND"
     // TODO - при реализации использовать логический оператор &&
     private <T> Predicate<T> and(Predicate<T> left, Predicate<T> right) {
-        throw new UnsupportedOperationException();
+        return left.and(right);
```
Comment (jemika) [inline src/test/java/lambda/part2/exercise/Exercise2.java]: also read TODO )

File: src/test/java/lambda/part2/exercise/Exercise2.java [hunk 3]
```diff
@@ -25,75 +25,91 @@ public void personHasNotEmptyLastNameAndFirstName() {
...
+        Predicate<Person> personHasEmptyFirstName = negateUsingLogicalOperator(p -> !p.getFirstName().isEmpty());
+        Predicate<Person> personHasEmptyLastName = negateUsingLogicalOperator(p -> !p.getLastName().isEmpty());
 
-        Predicate<Person> personHasNotEmptyFirstName = null;
-        Predicate<Person> personHasNotEmptyLastName = null;
+        Predicate<Person> personHasNotEmptyFirstName = negateUsingLogicalOperator(p -> p.getFirstName().isEmpty());
+        Predicate<Person> personHasNotEmptyLastName = negateUsingLogicalOperator(p -> p.getLastName().isEmpty());
 
-        Predicate<Person> personHasNotEmptyLastNameAndFirstName = null;
+        Predicate<Person> personHasNotEmptyLastNameAndFirstName = p -> !p.getLastName().isEmpty() && !p.getFirstName().isEmpty();
```
Comment (jemika) [inline src/test/java/lambda/part2/exercise/Exercise2.java]: u have written special function (andUsingLogicalOperator) for this condition, so why not use it ?)

Comment (Boochok) [reply src/test/java/lambda/part2/exercise/Exercise2.java]: because to this line I've already forgotten about that function)

**Comment by jemika** [comment]
approved!)

**Pull request merged**